File: finanze/infrastructure/repository/entity/entity_repository.py
from typing import Optional
from uuid import UUID

from application.ports.entity_port import EntityPort
from domain.entity import Entity
from infrastructure.repository.db.client import DBClient
from infrastructure.repository.entity.queries import EntityQueries


def _map_entity(row) -> Entity:
    return Entity(
        id=UUID(row["id"]),
        name=row["name"],
        natural_id=row["natural_id"],
        type=row["type"],
        origin=row["origin"],
        icon_url=row["icon_url"],
    )


class EntitySQLRepository(EntityPort):
    def __init__(self, client: DBClient):
        self._db_client = client

    def insert(self, entity: Entity):
        with self._db_client.tx() as cursor:
            cursor.execute(
                EntityQueries.INSERT,
                (
                    str(entity.id),
                    entity.name,
                    entity.natural_id,
                    entity.type,
                    entity.origin,
                    entity.icon_url,
                ),
            )

    def update(self, entity: Entity):
        with self._db_client.tx() as cursor:
            cursor.execute(
                EntityQueries.UPDATE,
                (
                    entity.name,
                    entity.natural_id,
                    entity.type,
                    entity.origin,
                    entity.icon_url,
                    str(entity.id),
                ),
            )

    def get_by_id(self, entity_id: UUID) -> Optional[Entity]:
        with self._db_client.read() as cursor:
            cursor.execute(EntityQueries.GET_BY_ID, (str(entity_id),))
            row = cursor.fetchone()
            if row:
                return _map_entity(row)
            return None

    def get_all(self) -> list[Entity]:
        with self._db_client.read() as cursor:
            cursor.execute(EntityQueries.GET_ALL)
            return [_map_entity(row) for row in cursor.fetchall()]

    def get_by_natural_id(self, natural_id: str) -> Optional[Entity]:
        with self._db_client.read() as cursor:
            cursor.execute(EntityQueries.GET_BY_NATURAL_ID, (natural_id,))
            row = cursor.fetchone()
            if row:
                return _map_entity(row)
            return None

    def get_by_name(self, name: str) -> Optional[Entity]:
        with self._db_client.read() as cursor:
            cursor.execute(EntityQueries.GET_BY_NAME, (name,))
            row = cursor.fetchone()
            if row:
                return _map_entity(row)
            return None

    def delete_by_id(self, entity_id: UUID):
        with self._db_client.tx() as cursor:
            cursor.execute(EntityQueries.DELETE_BY_ID, (entity_id,))

    def get_disabled_entities(self) -> list[Entity]:
        with self._db_client.read() as cursor:
            cursor.execute(EntityQueries.GET_DISABLED_ENTITIES)
            return [_map_entity(row) for row in cursor.fetchall()]

File: mobile/src/services/database/repositories/entityRepository.ts
import { DataManager } from "../dataManager"
import { Entity } from "@/domain"
import { EntityPort } from "@/application/ports"
import { EntityQueries } from "./queries"

export class EntityRepository implements EntityPort {
  constructor(private client: DataManager) {}

  async insert(entity: Entity): Promise<void> {
    await this.client.execute(EntityQueries.INSERT, [
      entity.id,
      entity.name,
      entity.naturalId ?? null,
      entity.type,
      entity.origin,
      entity.iconUrl ?? null,
    ])
  }

  async update(entity: Entity): Promise<void> {
    await this.client.execute(EntityQueries.UPDATE, [
      entity.name,
      entity.naturalId ?? null,
      entity.type,
      entity.origin,
      entity.iconUrl ?? null,
      entity.id,
    ])
  }

  async getAll(): Promise<Entity[]> {
    const result = await this.client.query<any>(EntityQueries.GET_ALL)
    return result.rows
      .map(this.mapEntity)
      .sort((a, b) => a.name.localeCompare(b.name))
  }

  async getById(id: string): Promise<Entity | null> {
    const result = await this.client.query<any>(EntityQueries.GET_BY_ID, [id])
    if (result.rows.length === 0) return null
    return this.mapEntity(result.rows[0])
  }

  async getByNaturalId(naturalId: string): Promise<Entity | null> {
    const result = await this.client.query<any>(
      EntityQueries.GET_BY_NATURAL_ID,
      [naturalId],
    )
    if (result.rows.length === 0) return null
    return this.mapEntity(result.rows[0])
  }

  async getByName(name: string): Promise<Entity | null> {
    const result = await this.client.query<any>(EntityQueries.GET_BY_NAME, [
      name,
    ])
    if (result.rows.length === 0) return null
    return this.mapEntity(result.rows[0])
  }

  async deleteById(entityId: string): Promise<void> {
    await this.client.execute(EntityQueries.DELETE_BY_ID, [entityId])
  }

  async getDisabledEntities(): Promise<Entity[]> {
    const result = await this.client.query<any>(
      EntityQueries.GET_DISABLED_ENTITIES,
    )

    return result.rows.map(this.mapEntity)
  }

  private mapEntity(row: any): Entity {
    return {
      id: row.id,
      name: row.name,
      naturalId: row.natural_id,
      type: row.type,
      origin: row.origin,
      iconUrl: row.icon_url,
    }
  }
}
